inst/training_202307/202307.qmd

```diff
@@ -0,0 +1,448 @@
+---
+title: "`verdepcheck`"
+author: "Pawel Rucki"
+date: 2023/08/07
+format: revealjs
+---
+
+## Agenda
+
+1. Motivation
+1. Solution
+1. How to use?
+1. How it works?
+1. Discussion
+1. Future work
+
+# Motivation
+
+```{.yaml code-line-numbers="7"}
+Type: Package
+Package: foo.package
+Version: 1.2.3.9000
+Depends:
+    R (>= 3.6)
+Imports:
+    dplyr,
+    (...)
+```
+
+The above specification indicates compatibility with **all** versions of `dplyr` which is very unlikely to be true.
+
+# Motivation
+
+```.default
+Error: object ‘foo’ is not exported by 'namespace:bar'
+```
+
+<br>
+
+```.default
+`foo()` was deprecated in bar 1.0.0.
+i Please use `baz()` instead.
+```
+
+<br>
+
+The new versions of dependencies might introduce a breaking changes.
+
+# Motivation
+
+R environments for clinical trials are usually not updated very frequently.
+
+Users might use `renv` for environment management and packages used might not get updates for some time.
+
+As a result, your package might be used with _older_ version of dependent packages.
+
+# Solution
+
+`verdepcheck` (**ver**sion of **dep**endencies **check**) - a tool for package developers to check your package against different set of dependencies. It is meant to be used primarily in CI.
+
+Available at:
+
+- [`insightsengineering/verdepcheck`](https://github.com/insightsengineering/verdepcheck)
+- [`insightsengineering/r-verdepcheck-action`](https://github.com/insightsengineering/r-verdepcheck-action)
+
+:::{.callout-important}
+Do not confuse `verdepcheck` with `revdepcheck` for reverse dependency checks. Those are totally different tools!
+:::
+
+#
+
+Let's call a "different set of dependencies" a _strategy_. Currently there are three _strategies_ implemented:
+
+- **max** - use _development_ version of dependencies
+- **release** - use _released_ version of dependencies (e.g. from CRAN)
+- **min** - use _minimal_ version of dependencies as per `DESCRIPTION` file
+
+This tool is looking into a **direct** dependencies only.
+
+## Explanation for direct dependencies
+
+The initial prototype of `min` strategy included recursive find of minimal dependency. This oftentimes lead into usage of already archived package or not able to compile very old package on modern systems architecture.
+
+In order to solve potential issues, package author would need to control / specify indirect dependencies which is counter intuitive.
+
+Therefore only **direct** dependencies are being looked into.
+
+# What are the benefits?
+
+* It forces you to specify minimal version of package dependencies to prevent incorrect package usage.
+* Incorporate information about upcoming breaking changes.
+
+# What's needed?
+
+For **max** strategy the algorithm will look for dependent packages references in a new section `Config/Needs/verdepcheck` of the `DESCRIPTION` file.
```

Comment:
in my opinion, this should be an argument, but thats probably because I designed switchr that way. 

If the manifest is something you pass the testing machinery, you can easily tell it to test new branches, etc, without *editing the DESCRIPTION file* which I hate the very concept of doing when the package has not changed. It changes the MD5, which means the package is different, even when its not different, and then you have package versions which have the same version number and different MD5s running around and its a huge awful  mess that didn't need to happen.

Reply:
I actually don't think it is going to be changed that frequently. A typical use that I am for-seeing would be that if once specified it would remain unchanged unless a new dependency is added / removed which will anyway involve changes of the `DESCRIPTION` file.
Having said that, `Config/Needs/*` is generally recommended place to store such metadata for CI that is also used by others.